google/cloud/spanner/client.h

```diff
@@ -133,7 +133,8 @@ class Client {
    */
   explicit Client(std::shared_ptr<Connection> conn, Options opts = {})
       : conn_(std::move(conn)),
-        opts_(spanner_internal::DefaultOptions(std::move(opts))) {}
+        opts_(spanner_internal::DefaultOptions(
+            internal::MergeOptions(std::move(opts), conn_->options()))) {}
```

Comment:
Ugh, I wrote athing and forgot to add it to my review. Starting over......

This logic is not tested. Should we add an `options()` method to `MockConnection`? or maybe have a new mock just for this purpose.

Then we can have tests like this:
```cc
/// verify that the connection options are merged before the client options are defaulted
TEST(ClientTest, UsesConnectionOptions) {
  auto const conn_opts = Options{}.set<GrpcNumChannelsOption>(42);
  auto mock = std::make_shared<MockConnection>();
  EXPECT_CALL(*mock, options).WillOnce(Return(conn_opts));
  EXPECT_CALL(*mock, Commit).WillOnce([]{
        auto opts = internal::CurrentOptions();
        EXPECT_EQ(42, opts.get<GrpcNumChannelsOption>());
      });

  auto client = Client(mock);
  (void)client.Commit(/* whatever goes here */);
}
```
I think we would want two other tests like this to check that:
1. client ctor options take precedence over connection ctor options
2. per-call options take precedence over client ctor options

Reply:
Also, now that I think about it, we technically don't need to Default the options here. (because the ones coming from the Connection must be full).

Reply:
Everyone should be able to merge in the default options without ill effect, but I agree that it is probably more structured if that is only done at a single point ... if only to stop ourselves asking the question each time we pass options around.

So, I removed the `DefaultOptions()` here, and propose that it **only** be used in `Make[Foo]Connection()`, with the result passed to the stub generator(s) and the `[Foo]ConnectionImpl` ctor.

Reply:
Done on the tests.

Reply:
Thanks